src/core/lib/gpr/arena.cc

```diff
@@ -118,35 +122,49 @@ void* gpr_arena_alloc(gpr_arena* arena, size_t size) {
   size = GPR_ROUND_UP_TO_ALIGNMENT_SIZE(size);
   size_t start = static_cast<size_t>(
       gpr_atm_no_barrier_fetch_add(&arena->size_so_far, size));
+  size_t total_size_of_arena_allocations = start + size;
   zone* z = &arena->initial_zone;
-  while (start > z->size_end) {
-    zone* next_z = (zone*)gpr_atm_acq_load(&z->next_atm);
-    if (next_z == nullptr) {
-      size_t next_z_size =
-          static_cast<size_t>(gpr_atm_no_barrier_load(&arena->size_so_far));
-      next_z = static_cast<zone*>(zalloc_aligned(
-          GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone)) + next_z_size));
-      next_z->size_begin = z->size_end;
-      next_z->size_end = z->size_end + next_z_size;
-      if (!gpr_atm_rel_cas(&z->next_atm, static_cast<gpr_atm>(NULL),
-                           (gpr_atm)next_z)) {
-        gpr_free_aligned(next_z);
-        next_z = (zone*)gpr_atm_acq_load(&z->next_atm);
+
+  // Check to see if the allocation isn't able to start in the initial zone.
+  // This statement is true only in the uncommon case because of our arena
+  // sizing historesis (that is, most calls should have a large enough initial
+  // zone and will not need to grow the arena).
+  if (start >= z->size_end) {
+    // Find the correct zone to start the allocation in
+    gpr_mu_lock(&arena->arena_growth_mutex);
+    while (start >= z->size_end) {
+      if (z->next == nullptr) {
+        size_t next_z_size = total_size_of_arena_allocations;
+        z->next = static_cast<zone*>(zalloc_aligned(
+            GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone)) + next_z_size));
+        z->next->size_begin = z->size_end;
+        z->next->size_end = z->size_end + next_z_size;
       }
+      z = z->next;
     }
-    z = next_z;
+    gpr_mu_unlock(&arena->arena_growth_mutex);
   }
-  if (start + size > z->size_end) {
+  // Now that we have a zone we could start in, can the allocation fit in this
+  // zone? If not, call arena_alloc again. Note that this means some space in
```

Comment:
I had the same thoughts and had tried a rewrite where we did not recurse and instead did this check inside the mutex. It didn't work. The problem is that we rely on size_so_far to tell us the next available memory to start an allocation at, so we unfortunately *need* the duplicate increment. 

Let's say we are trying to allocate x bytes onto the initial zone, which already has some bytes allocated on it. The remaining bytes available on the initial zone is less than x, so we decide that we should allocate it on the next zone. We go ahead and increment size_so_far only once and return the appropriate ptr to the available memory. Now someone calls gpr_arena_alloc and the ptr they receive, based on `size_so_far`, actually clobbers our x bytes. 

You can see that the inherent problem / feature here is that we don't use per-zone usage tracking, and instead just rely on this one total usage metric, size_so_far. I think it's actually a performance vs. space tradeoff to do it this way because I tried to write an implementation that uses per-zone tracking (avoids duplicate increment as well as the recursive call) and couldn't figure out how to make it equally performant. 

Another unfortunate thing about the duplicate increment is that we could theoretically waste a ton of space with a particularly antagonistic set of calls to gpr_arena_alloc.

Reply:
I see... So the fact that we re-increment `size_so_far` is the thing that prevents multiple simultaneous callers from stepping on each other.  In particular, I think (but please correct me if I'm wrong) that this means that whenever we need to allocate x bytes and there are x - n (where n > 0) remaining in the current zone, we will waste x bytes (x - n in the current zone and n in the new zone).

This behavior wasn't at all clear from first read of the code.  It would probably be a good idea to add this explanation into the comment.

Am I correct in saying that as soon as the arena needs a second zone, all subsequent callers will need to acquire the mutex, even if they don't wind up needing to add a new zone themselves (i.e., if the arena already has 2 zones, the next caller will need to acquire the mutex even if it fits in the second zone)?

Assuming that's correct, it seems like we could still avoid this recursion and thus avoid the need to release and then re-acquire the mutex:
- On lines 132 and 135, instead of checking whether the allocation can start in the current zone, check whether the allocation can end in the current zone.
- If we get to the last zone and still can't fit (the block on line 136), then just duplicate lines 123-125 to increment `size_so_far` before adding the new zone.

Also, given that we're acquiring the mutex anyway, it seems like we could avoid wasting space if we wanted by storing the amount of wasted space and the starting offset of the waste in fields protected by the mutex.  Then, callers that have to acquire the mutex anyway could use that information to grab some of the wasted space, if they fit.  Of course, this does add complexity, and the amount of waste is probably not huge, and this won't be an issue for most RPCs anyway, so it's probably not worth the complexity.

Reply:
Yes, I believe your statement on the precise number of wasted bytes is correct. I've added your explanation as a comment in the code.

Yes, once the arena needs a second zone, subsequent callers will need to acquire the mutex in order to navigate past the initial zone and find a zone they can potentially allocate in. 

I've changed the code to avoid recursion according to your suggestion.

We should be aware of the workaround for avoiding wasting space in case it's needed in the future, but I agree that it's probably not worth the added complexity.